components/raftstore-v2/src/operation/command/mod.rs

```diff
@@ -0,0 +1,165 @@
+            metrics.invalid_proposal.stale_command.inc();
+            return Err(e);
+        }
+        if let Err(mut e) = util::check_region_epoch(req, self.region(), true) {
+            if let Error::EpochNotMatch(_, new_regions) = &mut e {
+                // TODO: query sibling regions.
+                metrics.invalid_proposal.epoch_not_match.inc();
+            }
+            return Err(e);
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn propose_command<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let data = req.write_to_bytes().unwrap();
+        self.propose(ctx, data)
+    }
+
+    #[inline]
+    fn propose<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>, data: Vec<u8>) -> Result<u64> {
+        ctx.raft_metrics.propose.normal.inc();
+        PEER_PROPOSE_LOG_SIZE_HISTOGRAM.observe(data.len() as f64);
+        if data.len() as u64 > ctx.cfg.raft_entry_max_size.0 {
+            return Err(Error::RaftEntryTooLarge {
+                region_id: self.region_id(),
+                entry_size: data.len() as u64,
+            });
+        }
+        let last_index = self.raft_group().raft.raft_log.last_index();
+        self.raft_group_mut().propose(vec![], data)?;
+        if self.raft_group().raft.raft_log.last_index() == last_index {
+            // The message is dropped silently, this usually due to leader absence
+            // or transferring leader. Both cases can be considered as NotLeader error.
+            return Err(Error::NotLeader(self.region_id(), None));
+        }
+        Ok(last_index + 1)
+    }
+
+    #[inline]
+    fn enqueue_pending_proposal<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        mut proposal: Proposal<Vec<CmdResChannel>>,
+    ) {
+        let applied_to_current_term = self.applied_to_current_term();
+        if applied_to_current_term {
+            proposal.cb.notify_proposed();
+        }
+        proposal.must_pass_epoch_check = applied_to_current_term;
+        proposal.propose_time = Some(*ctx.current_time.get_or_insert_with(monotonic_raw_now));
+        self.proposals_mut().push(proposal);
+    }
+
+    #[inline]
+    pub fn schedule_apply_committed_entries<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        committed_entries: Vec<Entry>,
+    ) {
+        let last_entry = match committed_entries.last() {
+            Some(e) => e,
+            None => return,
+        };
+        let current_term = self.term();
+        let mut entry_and_proposals = vec![];
+        let queue = self.proposals_mut();
+        if !queue.is_empty() {
+            for e in committed_entries {
+                let mut proposal = queue.find_proposal(e.term, e.index, current_term);
+                if let Some(p) = &mut proposal && p.must_pass_epoch_check {
+                    // In this case the apply can be guaranteed to be successful. Invoke the
```

Comment:
why it can be guaranteed? Say what if right after notify_committed() the node is down?

Reply:
Then it will be applied after being restarted.